### DISCUSSION TITLE
Easiest way to zero negative values for some tracers?

### DISCUSSION THREAD

**Discussion by iuryt** [CliMA/Oceananigans.jl#2409] [Computational science!] [answered]
This a spin off from #2385 but I am creating a new discussion because I think that this is useful for other situations.
In my case, nutrients are consumed and sometimes could reach negative values. 
Despite being O(1e-5) negative values, I would like to zero them at each timestep.

I am also open for other alternatives to avoid these negative values on a NP model.

**Comment by glwagner** [comment, marked as answer]
Maybe

```julia
zero_tracer(sim) = map!(c -> ifelse(c < 0, 0.0, c), parent(sim.model.tracers.c), parent(sim.model.tracers.c))
simulation.callbacks[:zero_tracer] = Callback(zero_tracer)
```

**Reply by glwagner** [reply]
This might work too

```julia
zero_tracer(sim) = parent(c) .= max.(0, parent(c))
```

**Reply by iuryt** [reply]
```julia
zero_tracer(sim) = parent(sim.model.tracers.c) .= max.(0, parent(sim.model.tracers.c))
```
Do I need to specify that `c` comes from `model.tracers`?

**Reply by glwagner** [reply]
Well yes, `c` must be defined somehow, sorry, I just wrote it in shorthand to save space and make it clearer. You can also write

```julia
c = model.tracers.c
zero_tracer(sim) = parent(c) .= max.(0, parent(c))
```

**Reply by glwagner** [reply]
In the context of a model that can't handle negative values you can also write things like `max(0, c)` inside your kernels. That might be slightly more robust (and also avoids the extra callback computation).

**Reply by iuryt** [reply]
That would be great! How should I do that? By now I only have a kernel for computing mld and average light.

**Reply by iuryt** [reply]
> Well yes, `c` must be defined somehow, sorry, I just wrote it in shorthand to save space and make it clearer. You can also write
> 
> ```julia
> c = model.tracers.c
> zero_tracer(sim) = parent(c) .= max.(0, parent(c))
> ```

This works. Just tested.
